File: core/src/main/java/com/bindglam/utility/database/SQLiteDatabase.java
package com.bindglam.utility.database;

import org.bukkit.configuration.ConfigurationSection;
import org.jetbrains.annotations.NotNull;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class SQLiteDatabase implements Database {
    private static final int VALID_TIMEOUT = 500;

    private Connection connection;

    @Override
    public void connect(ConfigurationSection config) {
        try {
            Class.forName("org.sqlite.JDBC");
        } catch (ClassNotFoundException e) {
            throw new RuntimeException("Failed to load database driver", e);
        }

        try {
            createConnection();
        } catch (SQLException e) {
            throw new RuntimeException("Failed to load database", e);
        }
    }

    @Override
    public void close() {
        try {
            if (connection != null) {
                connection.close();
            }
        } catch (SQLException e) {
            throw new RuntimeException("Failed to close database", e);
        } finally {
            connection = null;
        }
    }

    @Override
    public @NotNull Connection getConnection() {
        try {
            if(connection == null || connection.isValid(VALID_TIMEOUT)) {
                close();
                createConnection();
            }
        } catch (SQLException e) {
            throw new RuntimeException("Failed to ensure connection", e);
        }
        return connection;
    }

    @Override
    public void evictConnection(@NotNull Connection connection) {
    }

    private void createConnection() throws SQLException {
        connection = DriverManager.getConnection("jdbc:sqlite:plugins/BindglamUtility/database.db");
    }
}

File: core/src/main/kotlin/com/bindglam/utility/database/SQLiteDatabase.kt
package com.bindglam.utility.database

import org.bukkit.configuration.ConfigurationSection
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException

class SQLiteDatabase : Database {
    private var connection: Connection? = null

    init {
        try {
            Class.forName("org.sqlite.JDBC")
        } catch (e: ClassNotFoundException) {
            throw RuntimeException("Failed to load database driver", e)
        }
    }

    // config is not used
    override fun connect(config: ConfigurationSection?) {
        try {
            connection = DriverManager.getConnection("jdbc:sqlite:plugins/BindglamUtility/database.db")
        } catch (e: SQLException) {
            throw RuntimeException("Failed to load database", e)
        }
    }

    override fun close() {
        try {
            if (connection != null) {
                connection!!.close()
            }
        } catch (e: SQLException) {
            throw RuntimeException("Failed to close database", e)
        } finally {
            connection = null
        }
    }

    override fun getConnection(): Connection {
        try {
            if (connection == null || connection?.isValid(VALID_TIMEOUT) == false) {
                close()
                connect(null)
            }
        } catch (e: SQLException) {
            throw RuntimeException("Failed to ensure connection", e)
        }
        return connection!!
    }

    override fun evictConnection(connection: Connection) {
    }

    companion object {
        private const val VALID_TIMEOUT = 500
    }
}